### PR TITLE
datetime serialization performance

### LINKED ISSUE #256: Serialization performance of datetime types
Currently, datetime type serialization is slow because it incurs heap allocations, even though the interface would allow them not to heap-alloc.

This should be changed by using predetermined buffer sizes and `std::format_to` instead of `std::format`.

### PR DESCRIPTION
**Pull request by mcb5637**
close #256 

nanobench results (bigger % number is better)

| relative |               ns/op |                op/s |    err% |     total | benchmark
|---------:|--------------------:|--------------------:|--------:|----------:|:----------
|   100.0% |              207.21 |        4,826,107.92 |    0.7% |      0.01 | `baseline`
|   134.1% |              154.57 |        6,469,472.81 |    0.5% |      0.01 | `buffered`
|   112.1% |              184.77 |        5,412,075.52 |    1.1% |      0.01 | `direct write`

### CODE REVIEW

File: src/rdf4cpp/writer/BufWriter.hpp [hunk 1]
```diff
@@ -366,6 +366,38 @@ struct BufOutputIteratorWriter : BufWriterBase<BufOutputIteratorWriter<OutIter>,
...
+    }
+
+    BufWriterOutputIterator& operator++() {
+
+        //*writer.cursor->data() = buff;
+        writer.cursor->advance(1);
+        return *this;
+    }
+    BufWriterOutputIterator operator++(int) {
+        assert(false);
```
Comment (liss-h) [inline src/rdf4cpp/writer/BufWriter.hpp]: Why is this assert here?

File: src/rdf4cpp/writer/BufWriter.hpp [hunk 2]
```diff
@@ -366,6 +366,38 @@ struct BufOutputIteratorWriter : BufWriterBase<BufOutputIteratorWriter<OutIter>,
...
+                write_ok = false;
+                assert(false);
+            }
+        }
+        return *writer.cursor->data();
+    }
+
+    BufWriterOutputIterator& operator++() {
+
+        //*writer.cursor->data() = buff;
```
Comment (liss-h) [inline src/rdf4cpp/writer/BufWriter.hpp]: remove comment

File: src/rdf4cpp/Timezone.hpp [hunk 3]
```diff
@@ -64,10 +64,37 @@ struct Timezone {
...
+        return buf;
+    }
+
+    [[nodiscard]] bool to_canonical_string(writer::BufWriterParts writer) const noexcept {
+        if (offset == std::chrono::minutes{0}) {
+            return writer::write_str("Z", writer);
+        }
+        auto h = std::chrono::floor<std::chrono::hours>(std::chrono::abs(offset));
+        auto m = std::chrono::abs(offset) - h;
+        return std::format_to(writer::BufWriterOutputIterator{writer}, "{}{:02}:{:02}", offset >= std::chrono::minutes{0} ? '+' : '-', h.count(), m.count()).write_ok;
```
Comment (liss-h) [inline src/rdf4cpp/Timezone.hpp]: Why does this use the adaptor instead of a buffer? Is this function used?

File: src/rdf4cpp/Timezone.hpp [hunk 4]
```diff
@@ -64,10 +64,37 @@ struct Timezone {
     }
 
     [[nodiscard]] std::string to_canonical_string() const noexcept {
-        if (offset.count() == 0)
-            return "Z";
-        auto c = offset.count();
-        return std::format("{:+03}:{:02}", c / 60, std::abs(c) % 60);
+        std::string buf;
+        writer::StringWriter ser{buf};
```
Comment (liss-h) [inline src/rdf4cpp/Timezone.hpp]: Use `StringWriter::oneshot`

File: src/rdf4cpp/datatypes/xsd/time/DateTime.cpp [hunk 5]
```diff
@@ -46,11 +46,16 @@ capabilities::Default<xsd_dateTime>::cpp_type capabilities::Default<xsd_dateTime
 
 template<>
 bool capabilities::Default<xsd_dateTime>::serialize_canonical_string(cpp_type const &value, writer::BufWriterParts writer) noexcept {
-    auto str = std::format("{:%Y-%m-%dT%H:%M:%S}", value.first);
-    if (value.second.has_value())
-        str += value.second->to_canonical_string();
-
-    return writer::write_str(str, writer);
+    //year,-,month,-,day,T,hours,:,min,:,sec,tz
+    std::array<char, 6+1+2+1+2 +1+ 2+1+2+1+6 + Timezone::max_canonical_string_chars> buff;
```
Comment (liss-h) [inline src/rdf4cpp/datatypes/xsd/time/DateTime.cpp]: Is seconds +6 because of subseconds (`xx.yyy`)? If yes, please state that in the comment

File: src/rdf4cpp/datatypes/xsd/time/DateTimeStamp.cpp [hunk 6]
```diff
@@ -51,8 +51,14 @@ capabilities::Default<xsd_dateTimeStamp>::cpp_type capabilities::Default<xsd_dat
 
 template<>
 bool capabilities::Default<xsd_dateTimeStamp>::serialize_canonical_string(cpp_type const &value, writer::BufWriterParts writer) noexcept {
-    auto str = std::format("{:%Y-%m-%dT%H:%M:%S%Z}", value);
-    return writer::write_str(str, writer);
+    //year,-,month,-,day,T,hours,:,min,:,sec,tz
+    std::array<char, 6+1+2+1+2 +1+ 2+1+2+1+6 + Timezone::max_canonical_string_chars> buff;
```
Comment (liss-h) [inline src/rdf4cpp/datatypes/xsd/time/DateTimeStamp.cpp]: same here (sec -> +6)

File: src/rdf4cpp/datatypes/xsd/time/Month.cpp [hunk 7]
```diff
@@ -25,11 +25,15 @@ capabilities::Default<xsd_gMonth>::cpp_type capabilities::Default<xsd_gMonth>::f
 
 template<>
 bool capabilities::Default<xsd_gMonth>::serialize_canonical_string(cpp_type const &value, writer::BufWriterParts writer) noexcept {
-    auto str = std::format("--{:%m}", value.first);
-    if (value.second.has_value())
-        str += value.second->to_canonical_string();
-
-    return writer::write_str(str, writer);
+    //assumes month is in [1,12]
```
Comment (liss-h) [inline src/rdf4cpp/datatypes/xsd/time/Month.cpp]: Can month be larger than this interval?

File: src/rdf4cpp/writer/BufWriter.hpp [hunk 8]
```diff
@@ -225,7 +225,7 @@ struct StringWriter : BufWriterBase<StringWriter, StringBuffer> {
     static void flush_impl(Buffer &buffer, Cursor &cursor, size_t additional_cap) noexcept {
         auto const bytes_filled = cursor.data() - buffer.buffer_->data();
 
-        buffer.buffer_->resize(std::bit_ceil(buffer.buffer_->size() + additional_cap));
+        buffer.buffer_->resize(std::bit_ceil(std::max(buffer.buffer_->size() + additional_cap, buffer.buffer_->size() * 2)));
```
Comment (liss-h) [inline src/rdf4cpp/writer/BufWriter.hpp]: Why this change?

File: tests/datatype/tests_time_types.cpp [hunk 9]
```diff
@@ -611,3 +615,83 @@ TEST_CASE("arithmetic") {
...
+    if (!std::format_to(rdf4cpp::writer::BufWriterOutputIterator{writer}, "{:%Y-%m-%d}", value.first).write_ok) {
+        return false;
+    }
+    if (value.second.has_value()) {
+        return value.second->to_canonical_string(writer);
+    }
+    return true;
+}
+
+TEST_CASE("serialization perf") {
```
Comment (liss-h) [inline tests/datatype/tests_time_types.cpp]: I would not merge the benchmark itself, we measured it and know the fastest solution now. Benchmark related stuff can be removed

File: src/rdf4cpp/datatypes/registry/util/DateTimeUtils.hpp [hunk 10]
```diff
@@ -142,6 +142,17 @@ inline std::optional<std::chrono::milliseconds> parse_duration_milliseconds(std:
     return parse_milliseconds<datatype>(res_s);
 }
 
+inline char* canonical_seconds_remove_empty_millis(char* it) {
+    for (size_t m = 0; m<3; ++m) {
+        if (it[-1] != '0')
```
Comment (liss-h) [inline src/rdf4cpp/datatypes/registry/util/DateTimeUtils.hpp]: Please don't use array syntax for that. `*(it - 1)` is clearer

File: src/rdf4cpp/datatypes/registry/util/DateTimeUtils.hpp [hunk 11]
```diff
@@ -142,6 +142,17 @@ inline std::optional<std::chrono::milliseconds> parse_duration_milliseconds(std:
     return parse_milliseconds<datatype>(res_s);
 }
 
+inline char* canonical_seconds_remove_empty_millis(char* it) {
```
Comment (liss-h) [inline src/rdf4cpp/datatypes/registry/util/DateTimeUtils.hpp]: `char *conanical_..`

**Review by liss-h — CHANGES_REQUESTED** [review]
Thx. Some comments below

File: src/rdf4cpp/writer/BufWriter.hpp (hunk 1 above)
Comment (mcb5637) [reply src/rdf4cpp/writer/BufWriter.hpp]: sorry, i forgot to remove the entire class, its unused. (leftover from direct write).

File: src/rdf4cpp/datatypes/xsd/time/Month.cpp (hunk 7 above)
Comment (mcb5637) [reply src/rdf4cpp/datatypes/xsd/time/Month.cpp]: std::chrono::month is [0,255], but from_string checks, that it is in [1,12]

File: src/rdf4cpp/writer/BufWriter.hpp (hunk 8 above)
Comment (mcb5637) [reply src/rdf4cpp/writer/BufWriter.hpp]: also a leftover i forgot to remove

File: src/rdf4cpp/datatypes/xsd/time/Month.cpp (hunk 7 above)
Comment (liss-h) [reply src/rdf4cpp/datatypes/xsd/time/Month.cpp]: I would just add +1 for safety then. I don't think this can happen but better safe than sorry

File: src/rdf4cpp/Timezone.hpp [hunk 12]
```diff
@@ -63,11 +63,25 @@ struct Timezone {
         return tz;
     }
 
+    // sign, hours, :, minutes
+    // assumes offset stays inside spec (+-14 hours)
```
Comment (liss-h) [inline src/rdf4cpp/Timezone.hpp]: Same thing here, could this be larger?

File: tests/CMakeLists.txt [hunk 13]
```diff
@@ -324,6 +324,7 @@ add_test(NAME tests_NodeStorage_specialization COMMAND tests_NodeStorage_special
 add_executable(tests_time_types datatype/tests_time_types.cpp)
 target_link_libraries(tests_time_types
         doctest::doctest
+        nanobench::nanobench
```
Comment (liss-h) [inline tests/CMakeLists.txt]: leftover?

File: tests/datatype/tests_time_types.cpp [hunk 14]
```diff
@@ -5,6 +5,10 @@
 #include <rdf4cpp.hpp>
 #include <rdf4cpp/storage/reference_node_storage/SyncReferenceNodeStorage.hpp>
 
+#define ANKERL_NANOBENCH_IMPLEMENT
+#include <nanobench.h>
+#include <random>
```
Comment (liss-h) [inline tests/datatype/tests_time_types.cpp]: leftover?

**Review by liss-h — CHANGES_REQUESTED** [review]
Just some minor things left

File: src/rdf4cpp/Timezone.hpp (hunk 12 above)
Comment (mcb5637) [reply src/rdf4cpp/Timezone.hpp]: the next bigger guess would be `std::numeric_limits<int64_t>::digits10+1`

Comment (liss-h) [reply src/rdf4cpp/Timezone.hpp]: General question: Does chrono stuff not enforce the limits of the datatypes? This would then mean that all buffer calculations would actually need to be based on the underlying datatypes and not on their logical size, right?

**Review by liss-h — CHANGES_REQUESTED** [review]
The buffer size change was meant for every (datetime) type

**Comment by mcb5637** [comment]
Should be all. Year is limited to `[-32767, 32767]` (6 chars) Month and Day to `[​0​, 255]` (3 chars).

File: src/rdf4cpp/datatypes/registry/util/DateTimeUtils.hpp [hunk 15]
```diff
@@ -277,6 +288,51 @@ inline std::chrono::year_month_day normalize(std::chrono::year_month_day i) {
...
+    } else {
+        return 1 + number_of_digits(num / base);
+    }
+}
+static_assert(number_of_digits(0)==1);
+static_assert(number_of_digits(9)==1);
+static_assert(number_of_digits(-1)==2);
+static_assert(number_of_digits(10)==2);
+static_assert(number_of_digits(std::numeric_limits<uint64_t>::max())==std::numeric_limits<uint64_t>::digits10+1);
+struct chrono_max_canonical_string_chars {
```
Comment (liss-h) [inline src/rdf4cpp/datatypes/registry/util/DateTimeUtils.hpp]: should be a namespace instead with `inline constexpr` variables

File: src/rdf4cpp/datatypes/registry/util/DateTimeUtils.hpp [hunk 16]
```diff
@@ -277,6 +288,51 @@ inline std::chrono::year_month_day normalize(std::chrono::year_month_day i) {
...
+    //std::chrono::year is in [-32767, 32767]
+    static constexpr size_t year = 6;
+    static_assert(std::chrono::year::min() == std::chrono::year(-32767));  //NOLINT
+    static_assert(std::chrono::year::max() == std::chrono::year(32767));   //NOLINT
```
Comment (liss-h) [inline src/rdf4cpp/datatypes/registry/util/DateTimeUtils.hpp:310]: why is the year digits not just the digits in the max value + 1? (or max of number of digits in min and max)

File: src/rdf4cpp/datatypes/registry/util/DateTimeUtils.hpp [hunk 17]
```diff
@@ -277,6 +288,51 @@ inline std::chrono::year_month_day normalize(std::chrono::year_month_day i) {
...
+    //std::chrono::day is in [0, 255]
+    static constexpr size_t day = 3;
```
Comment (liss-h) [inline src/rdf4cpp/datatypes/registry/util/DateTimeUtils.hpp]: same here

File: src/rdf4cpp/datatypes/registry/util/DateTimeUtils.hpp [hunk 18]
```diff
@@ -277,6 +288,51 @@ inline std::chrono::year_month_day normalize(std::chrono::year_month_day i) {
...
+    //std::chrono::month is in [0, 255]
+    static constexpr size_t month = 3;
```
Comment (liss-h) [inline src/rdf4cpp/datatypes/registry/util/DateTimeUtils.hpp]: same here

**Review by liss-h — CHANGES_REQUESTED** [review]

File: src/rdf4cpp/datatypes/registry/util/DateTimeUtils.hpp (hunk 17 above)
Comment (mcb5637) [reply src/rdf4cpp/datatypes/registry/util/DateTimeUtils.hpp]: day/month do not have max defined

Comment (liss-h) [reply src/rdf4cpp/datatypes/registry/util/DateTimeUtils.hpp]: I just looked at cpp reference, they don't even have an `underlying_type` typedef defined.. great. Ok then you can leave it as it is and we just hope it doesn't cause issues

**Review by liss-h — APPROVED** [review]
